### PR TITLE
Tests for deploy_sandbox_shared_setup

### PR DESCRIPTION
**Pull request by jessesnyder**
## Description
Test coverage and refactoring of this complex CLI helper function and its service code

## Motivation and Context
* improve coverage
* Improve code quality safely

### CODE REVIEW

**Comment by jessesnyder** [comment]
@suchow I have a test passing locally which relies on the fact that I'm logged into heroku through the heroku CLI. I'm not exactly sure how we'd replicate this on Travis; maybe it's not possible(?) 

Now that I have an end-to-end test running locally, I can get pretty far by extracting the Heroku interactions and then replacing them with mocks in tests, but at least one full-stack "smoke test" always feels good.

**Comment by suchow** [comment]
@jessesnyder One possibility is that we can create a test Heroku account that's part of the Dallinger account, and then add the user name and password as environment variables either using the [Travis settings page](https://travis-ci.org/Dallinger/Dallinger/settings) or via `travis encrypt` in `.travis.yml`. Then, your test can use `pexpect` to log in to the test account by retrieving the credentials from the environment variables. Whew!

**Comment by jessesnyder** [comment]
@suchow Yes, still in progress. I'm on vacation until Thursday, but will probably poke at it a bit before then. I'll compile a list of things I'm touching that should be tested manually (`hibernate` and `awaken` will be among them).

**Comment by suchow** [comment]
@jessesnyder sounds good, i'll wait for further instruction.

**Comment by jessesnyder** [comment]
@suchow OK, I think this is enough for one PR. :-)

Worth manual testing:

- command that calls `deploy_sandbox_shared_setup()`
- `destroy` CLI command
- `hibernate` CLI command
- `awaken` CLI command
- `logs` CLI command
- `monitor` CLI command
- `data.backup()`
- `data.copy_heroku_to_local()`

File: dallinger/command_line.py [hunk 1]
```diff
@@ -501,74 +472,63 @@ def deploy_sandbox_shared_setup(verbose=True, app=None, web_procs=1, exp_config=
...
-    db_url = subprocess.check_output([
-        "heroku", "config:get", "DATABASE_URL", "--app", app_name(id)
-    ])
-
     # Set the notification URL and database URL in the config file.
     config.extend({
-        "notification_url": u"http://" + app_name(id) + ".herokuapp.com/notifications",
-        "database_url": db_url.rstrip().decode('utf8'),
+        "notification_url": heroku_app.url + "/notifications",
+        "database_url": heroku_app.db_url,
```
Comment (vlall) [inline dallinger/command_line.py]: For `dallinger sandbox --verbose`. I receive a TypeError. I just casted it to unicode and fixed it. I'll push after making sure there aren't anymore TypeErrors.

❯❯ Saving the URL of the postgres database...
Waiting for database postgresql-elliptical-25308... available
Traceback (most recent call last):
  File "/Users/vlall/.virtualenvs/d2/bin/dallinger", line 11, in <module>
    load_entry_point('dallinger', 'console_scripts', 'dallinger')()
  File "/Users/vlall/.virtualenvs/d2/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/vlall/.virtualenvs/d2/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/vlall/.virtualenvs/d2/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/vlall/.virtualenvs/d2/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/vlall/.virtualenvs/d2/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/vlall/prog/dev/Dallinger/dallinger/command_line.py", line 557, in sandbox
    _deploy_in_mode(u'sandbox', app, verbose)
  File "/Users/vlall/prog/dev/Dallinger/dallinger/command_line.py", line 549, in _deploy_in_mode
    deploy_sandbox_shared_setup(verbose=verbose, app=app)
  File "/Users/vlall/prog/dev/Dallinger/dallinger/command_line.py", line 493, in deploy_sandbox_shared_setup
    "database_url": heroku_app.db_url,
  File "/Users/vlall/prog/dev/Dallinger/dallinger/config.py", line 126, in extend
    expected_type=expected_type,
TypeError: Got 'https://dlgr-5187084d.herokuapp.com//notifications' for notification_url, expected <type 'unicode'>

**Comment by vlall** [comment]
I ran the app, which failed, so I checked the heroku logs. I think perhaps the double back slashes caused a parameter error, will try again without the slashes.

<SetHITTypeNotificationResponse><OperationRequest><RequestId>155da505-e952-4fcf-a2ee-aac8b3e83211</RequestId></OperationRequest><SetHITTypeNotificationResult><Request><IsValid>False</IsValid><Errors><Error><Code>AWS.MechanicalTurk.InvalidParameterValue</Code><Message>The value "https://dlgr-aff1a7f3.herokuapp.com//notifications" is not valid for the parameter Destination. (1503686284827 s)</Message><Data><Key>Parameter</Key><Value>Destination</Value></Data><Data><Key>Value</Key><Value>https://dlgr-aff1a7f3.herokuapp.com//notifications</Value></Data><Data><Key>Parameter</Key><Value>Destination</Value></Data><Data><Key>Value</Key><Value>https://dlgr-aff1a7f3.herokuapp.com//notifications</Value></Data></Error></Errors></Request></SetHITTypeNotificationResult></SetHITTypeNotificationResponse>

File: dallinger/command_line.py (hunk 1 above)
Comment (jessesnyder) [reply dallinger/command_line.py]: @vlall Just checked my integration test, and that fails for the same reason. I think the properties and return values from HerokuApp methods should probably be changed to return unicode objects.

**Comment by jessesnyder** [comment]
@vlall I think it makes sense to not include a trailing slash on HerokuApp's `url` property, and have client code that uses it to construct longer urls add the slash. I've made that change, but won't push since it sounds like you may have one or more commits to push.

**Pull request merged**